// tamejs/test/cases/17.out.js
var tame = require('tamejs').runtime;
var __tame_defer_cb = null;
var __tame_fn_0 = function (__tame_k) {
    tame.setActiveCb (__tame_defer_cb);
    "use strict" ;
    
    var assert = require ( "assert" ) ;
    
    var res = 0 ;
    var i = 0 ;
    function foo (i, cb) {
        var __tame_defer_cb = tame.findDeferCb ([i, cb]);
        tame.setActiveCb (__tame_defer_cb);
        var __tame_this = this;
        var __tame_fn_1 = function (__tame_k) {
            tame.setActiveCb (__tame_defer_cb);
            var __tame_fn_2 = function (__tame_k) {
                tame.setActiveCb (__tame_defer_cb);
                var __tame_defers = new tame.Deferrals (__tame_k);
                var __tame_fn_3 = function (__tame_k) {
                    tame.setActiveCb (__tame_defer_cb);
                    setTimeout (
                    __tame_defers.defer ( { 
                        func_name : "foo",
                        parent_cb : __tame_defer_cb,
                        line : 11,
                        file : "test/cases/17.tjs"
                    } )
                    , i ) ;
                    tame.callChain([__tame_k]);
                    tame.setActiveCb (null);
                };
                __tame_fn_3(tame.end);
                __tame_defers._fulfill();
                tame.setActiveCb (null);
            };
            var __tame_fn_4 = function (__tame_k) {
                tame.setActiveCb (__tame_defer_cb);
                cb ( i ) ;
                tame.callChain([__tame_k]);
                tame.setActiveCb (null);
            };
            tame.callChain([__tame_fn_2, __tame_fn_4, __tame_k]);
            tame.setActiveCb (null);
        };
        tame.callChain([__tame_fn_1, __tame_k]);
        tame.setActiveCb (null);
    }
    ;
    
    
    var ok = true ;
    
    exports [ "a test of fulfill values" ] = {
    
    run :
    function  (cb) {
        var __tame_defer_cb = tame.findDeferCb ([cb]);
        tame.setActiveCb (__tame_defer_cb);
        var __tame_this = this;
        var __tame_fn_5 = function (__tame_k) {
            tame.setActiveCb (__tame_defer_cb);
            var __tame_fn_6 = function (__tame_k) {
                tame.setActiveCb (__tame_defer_cb);
                var __tame_k_implicit = {};
                var i = 0;
                var __tame_fn_7 = function (__tame_k) {
                    tame.setActiveCb (__tame_defer_cb);
                    var __tame_fn_8 = function (__tame_k) {
                        tame.setActiveCb (__tame_defer_cb);
                        i ++
                        tame.callChain([__tame_fn_7, __tame_k]);
                        tame.setActiveCb (null);
                    };
                    __tame_k_implicit.k_break = __tame_k;
                    __tame_k_implicit.k_continue = function() { __tame_fn_8(__tame_k); };
                    if (i < 20) {
                        var __tame_fn_14 = function (__tame_k) {
                            tame.setActiveCb (__tame_defer_cb);
                            var x = 3 * Math . random ( ) ;
                            var __tame_fn_9 = function (__tame_k) {
                                tame.setActiveCb (__tame_defer_cb);
                                var y, z;
                                var __tame_fn_10 = function (__tame_k) {
                                    tame.setActiveCb (__tame_defer_cb);
                                    var __tame_defers = new tame.Deferrals (__tame_k);
                                    var __tame_fn_11 = function (__tame_k) {
                                        tame.setActiveCb (__tame_defer_cb);
                                        foo ( x ,
                                        __tame_defers.defer ( { 
                                            assign_fn : 
                                                function () {
                                                    y = arguments[0];
                                                    z = arguments[1];
                                                }
                                                ,
                                            parent_cb : __tame_defer_cb,
                                            line : 24,
                                            file : "test/cases/17.tjs"
                                        } )
                                        ) ;
                                        tame.callChain([__tame_k]);
                                        tame.setActiveCb (null);
                                    };
                                    __tame_fn_11(tame.end);
                                    __tame_defers._fulfill();
                                    tame.setActiveCb (null);
                                };
                                var __tame_fn_12 = function (__tame_k) {
                                    tame.setActiveCb (__tame_defer_cb);
                                    var __tame_fn_13 = function (__tame_k) {
                                        tame.setActiveCb (__tame_defer_cb);
                                        ok = false ;
                                        tame.callChain([__tame_k]);
                                        tame.setActiveCb (null);
                                    };
                                    if (x != y) {
                                        tame.callChain([__tame_fn_13, __tame_k]);
                                    } else {
                                        tame.callChain([__tame_k]);
                                    }
                                    tame.setActiveCb (null);
                                };
                                tame.callChain([__tame_fn_10, __tame_fn_12, __tame_k]);
                                tame.setActiveCb (null);
                            };
                            tame.callChain([__tame_fn_9, __tame_k]);
                            tame.setActiveCb (null);
                        };
                        tame.callChain([__tame_fn_14, __tame_fn_8, __tame_k]);
                    } else {
                        tame.callChain([__tame_k]);
                    }
                    tame.setActiveCb (null);
                };
                tame.callChain([__tame_fn_7, __tame_k]);
                tame.setActiveCb (null);
            };
            var __tame_fn_15 = function (__tame_k) {
                tame.setActiveCb (__tame_defer_cb);
                var __tame_k_implicit = {};
                var i = 0;
                var __tame_fn_16 = function (__tame_k) {
                    tame.setActiveCb (__tame_defer_cb);
                    var __tame_fn_17 = function (__tame_k) {
                        tame.setActiveCb (__tame_defer_cb);
                        i ++
                        tame.callChain([__tame_fn_16, __tame_k]);
                        tame.setActiveCb (null);
                    };
                    __tame_k_implicit.k_break = __tame_k;
                    __tame_k_implicit.k_continue = function() { __tame_fn_17(__tame_k); };
                    if (i < 20) {
                        var __tame_fn_23 = function (__tame_k) {
                            tame.setActiveCb (__tame_defer_cb);
                            var x = 3 * Math . random ( ) ;
                            var __tame_fn_18 = function (__tame_k) {
                                tame.setActiveCb (__tame_defer_cb);
                                var res;
                                var __tame_fn_19 = function (__tame_k) {
                                    tame.setActiveCb (__tame_defer_cb);
                                    var __tame_defers = new tame.Deferrals (__tame_k);
                                    var __tame_fn_20 = function (__tame_k) {
                                        tame.setActiveCb (__tame_defer_cb);
                                        foo ( x ,
                                        __tame_defers.defer ( { 
                                            assign_fn : 
                                                function () {
                                                    res = arguments[0];
                                                }
                                                ,
                                            parent_cb : __tame_defer_cb,
                                            line : 30,
                                            file : "test/cases/17.tjs"
                                        } )
                                        ) ;
                                        tame.callChain([__tame_k]);
                                        tame.setActiveCb (null);
                                    };
                                    __tame_fn_20(tame.end);
                                    __tame_defers._fulfill();
                                    tame.setActiveCb (null);
                                };
                                var __tame_fn_21 = function (__tame_k) {
                                    tame.setActiveCb (__tame_defer_cb);
                                    var __tame_fn_22 = function (__tame_k) {
                                        tame.setActiveCb (__tame_defer_cb);
                                        ok = false ;
                                        tame.callChain([__tame_k]);
                                        tame.setActiveCb (null);
                                    };
                                    if (x != res) {
                                        tame.callChain([__tame_fn_22, __tame_k]);
                                    } else {
                                        tame.callChain([__tame_k]);
                                    }
                                    tame.setActiveCb (null);
                                };
                                tame.callChain([__tame_fn_19, __tame_fn_21, __tame_k]);
                                tame.setActiveCb (null);
                            };
                            tame.callChain([__tame_fn_18, __tame_k]);
                            tame.setActiveCb (null);
                        };
                        tame.callChain([__tame_fn_23, __tame_fn_17, __tame_k]);
                    } else {
                        tame.callChain([__tame_k]);
                    }
                    tame.setActiveCb (null);
                };
                tame.callChain([__tame_fn_16, __tame_k]);
                tame.setActiveCb (null);
            };
            var __tame_fn_24 = function (__tame_k) {
                tame.setActiveCb (__tame_defer_cb);
                cb ( ) ;
                tame.callChain([__tame_k]);
                tame.setActiveCb (null);
            };
            tame.callChain([__tame_fn_6, __tame_fn_15, __tame_fn_24, __tame_k]);
            tame.setActiveCb (null);
        };
        tame.callChain([__tame_fn_5, __tame_k]);
        tame.setActiveCb (null);
    }
    ,
    
    check :
    function  () {
        assert . ok ( ok , "fulfilled with proper values" ) ;
    }
    } ;
    tame.callChain([__tame_k]);
    tame.setActiveCb (null);
};
__tame_fn_0 (tame.end);
