// tamejs/test/cases/24.out.js
var tame = require('tamejs').runtime;
var __tame_defer_cb = null;
var __tame_fn_0 = function (__tame_k) {
    tame.setActiveCb (__tame_defer_cb);
    "use strict" ;
    
    var assert = require ( "assert" ) ;
    
    var results = [ ] ;
    function blah (ev) {
        var __tame_defer_cb = tame.findDeferCb ([ev]);
        tame.setActiveCb (__tame_defer_cb);
        var __tame_this = this;
        var __tame_fn_1 = function (__tame_k) {
            tame.setActiveCb (__tame_defer_cb);
            var __tame_fn_2 = function (__tame_k) {
                tame.setActiveCb (__tame_defer_cb);
                var __tame_defers = new tame.Deferrals (__tame_k);
                var __tame_fn_3 = function (__tame_k) {
                    tame.setActiveCb (__tame_defer_cb);
                    setTimeout (
                    __tame_defers.defer ( { 
                        func_name : "blah",
                        parent_cb : __tame_defer_cb,
                        line : 10,
                        file : "test/cases/24.tjs"
                    } )
                    , 10*Math . random ( ) ) ;
                    tame.callChain([__tame_k]);
                    tame.setActiveCb (null);
                };
                __tame_fn_3(tame.end);
                __tame_defers._fulfill();
                tame.setActiveCb (null);
            };
            var __tame_fn_4 = function (__tame_k) {
                tame.setActiveCb (__tame_defer_cb);
                ev ( ) ;
                tame.callChain([__tame_k]);
                tame.setActiveCb (null);
            };
            tame.callChain([__tame_fn_2, __tame_fn_4, __tame_k]);
            tame.setActiveCb (null);
        };
        tame.callChain([__tame_fn_1, __tame_k]);
        tame.setActiveCb (null);
    }
    ;
    function runsw (x, cb) {
        var __tame_defer_cb = tame.findDeferCb ([x, cb]);
        tame.setActiveCb (__tame_defer_cb);
        var __tame_this = this;
        var __tame_fn_5 = function (__tame_k) {
            tame.setActiveCb (__tame_defer_cb);
            var __tame_fn_6 = function (__tame_k) {
                tame.setActiveCb (__tame_defer_cb);
                var __tame_switch_x = x;
                var __tame_k_implicit = {};
                __tame_k_implicit.k_break = __tame_k;
                var __tame_fn_7 = function (__tame_k) {
                    tame.setActiveCb (__tame_defer_cb);
                    var __tame_fn_8 = function (__tame_k) {
                        tame.setActiveCb (__tame_defer_cb);
                        var __tame_defers = new tame.Deferrals (__tame_k);
                        var __tame_fn_9 = function (__tame_k) {
                            tame.setActiveCb (__tame_defer_cb);
                            blah (
                            __tame_defers.defer ( { 
                                func_name : "runsw",
                                parent_cb : __tame_defer_cb,
                                line : 18,
                                file : "test/cases/24.tjs"
                            } )
                            ) ;
                            tame.callChain([__tame_k]);
                            tame.setActiveCb (null);
                        };
                        __tame_fn_9(tame.end);
                        __tame_defers._fulfill();
                        tame.setActiveCb (null);
                    };
                    var __tame_fn_11 = function (__tame_k) {
                        tame.setActiveCb (__tame_defer_cb);
                        results . push ( 0 ) ;
                        var __tame_fn_10 = function (__tame_k) {
                            tame.setActiveCb (__tame_defer_cb);
                            __tame_k_implicit.k_break();
                            tame.setActiveCb (null);
                        };
                        tame.callChain([__tame_fn_10, __tame_k]);
                        tame.setActiveCb (null);
                    };
                    tame.callChain([__tame_fn_8, __tame_fn_11, __tame_k]);
                    tame.setActiveCb (null);
                };
                var __tame_fn_12 = function (__tame_k) {
                    tame.setActiveCb (__tame_defer_cb);
                    var __tame_fn_13 = function (__tame_k) {
                        tame.setActiveCb (__tame_defer_cb);
                        var __tame_defers = new tame.Deferrals (__tame_k);
                        var __tame_fn_14 = function (__tame_k) {
                            tame.setActiveCb (__tame_defer_cb);
                            blah (
                            __tame_defers.defer ( { 
                                func_name : "runsw",
                                parent_cb : __tame_defer_cb,
                                line : 23,
                                file : "test/cases/24.tjs"
                            } )
                            ) ;
                            tame.callChain([__tame_k]);
                            tame.setActiveCb (null);
                        };
                        __tame_fn_14(tame.end);
                        __tame_defers._fulfill();
                        tame.setActiveCb (null);
                    };
                    var __tame_fn_16 = function (__tame_k) {
                        tame.setActiveCb (__tame_defer_cb);
                        results . push ( 1 ) ;
                        var __tame_fn_15 = function (__tame_k) {
                            tame.setActiveCb (__tame_defer_cb);
                            __tame_k_implicit.k_break();
                            tame.setActiveCb (null);
                        };
                        tame.callChain([__tame_fn_15, __tame_k]);
                        tame.setActiveCb (null);
                    };
                    tame.callChain([__tame_fn_13, __tame_fn_16, __tame_k]);
                    tame.setActiveCb (null);
                };
                var __tame_fn_17 = function (__tame_k) {
                    tame.setActiveCb (__tame_defer_cb);
                    var __tame_fn_18 = function (__tame_k) {
                        tame.setActiveCb (__tame_defer_cb);
                        var __tame_defers = new tame.Deferrals (__tame_k);
                        var __tame_fn_19 = function (__tame_k) {
                            tame.setActiveCb (__tame_defer_cb);
                            blah (
                            __tame_defers.defer ( { 
                                func_name : "runsw",
                                parent_cb : __tame_defer_cb,
                                line : 27,
                                file : "test/cases/24.tjs"
                            } )
                            ) ;
                            tame.callChain([__tame_k]);
                            tame.setActiveCb (null);
                        };
                        __tame_fn_19(tame.end);
                        __tame_defers._fulfill();
                        tame.setActiveCb (null);
                    };
                    var __tame_fn_20 = function (__tame_k) {
                        tame.setActiveCb (__tame_defer_cb);
                        results . push ( 2 ) ;
                        tame.callChain([__tame_k]);
                        tame.setActiveCb (null);
                    };
                    tame.callChain([__tame_fn_18, __tame_fn_20, __tame_k]);
                    tame.setActiveCb (null);
                };
                var __tame_fn_21 = function (__tame_k) {
                    tame.setActiveCb (__tame_defer_cb);
                    var __tame_fn_22 = function (__tame_k) {
                        tame.setActiveCb (__tame_defer_cb);
                        var __tame_defers = new tame.Deferrals (__tame_k);
                        var __tame_fn_23 = function (__tame_k) {
                            tame.setActiveCb (__tame_defer_cb);
                            blah (
                            __tame_defers.defer ( { 
                                func_name : "runsw",
                                parent_cb : __tame_defer_cb,
                                line : 31,
                                file : "test/cases/24.tjs"
                            } )
                            ) ;
                            tame.callChain([__tame_k]);
                            tame.setActiveCb (null);
                        };
                        __tame_fn_23(tame.end);
                        __tame_defers._fulfill();
                        tame.setActiveCb (null);
                    };
                    var __tame_fn_24 = function (__tame_k) {
                        tame.setActiveCb (__tame_defer_cb);
                        results . push ( 3 ) ;
                        tame.callChain([__tame_k]);
                        tame.setActiveCb (null);
                    };
                    tame.callChain([__tame_fn_22, __tame_fn_24, __tame_k]);
                    tame.setActiveCb (null);
                };
                var __tame_fn_25 = function (__tame_k) {
                    tame.setActiveCb (__tame_defer_cb);
                    var __tame_fn_26 = function (__tame_k) {
                        tame.setActiveCb (__tame_defer_cb);
                        var __tame_defers = new tame.Deferrals (__tame_k);
                        var __tame_fn_27 = function (__tame_k) {
                            tame.setActiveCb (__tame_defer_cb);
                            blah (
                            __tame_defers.defer ( { 
                                func_name : "runsw",
                                parent_cb : __tame_defer_cb,
                                line : 34,
                                file : "test/cases/24.tjs"
                            } )
                            ) ;
                            tame.callChain([__tame_k]);
                            tame.setActiveCb (null);
                        };
                        __tame_fn_27(tame.end);
                        __tame_defers._fulfill();
                        tame.setActiveCb (null);
                    };
                    var __tame_fn_29 = function (__tame_k) {
                        tame.setActiveCb (__tame_defer_cb);
                        results . push ( 4 ) ;
                        var __tame_fn_28 = function (__tame_k) {
                            tame.setActiveCb (__tame_defer_cb);
                            __tame_k_implicit.k_break();
                            tame.setActiveCb (null);
                        };
                        tame.callChain([__tame_fn_28, __tame_k]);
                        tame.setActiveCb (null);
                    };
                    tame.callChain([__tame_fn_26, __tame_fn_29, __tame_k]);
                    tame.setActiveCb (null);
                };
                var __tame_switch_calls = [__tame_fn_7, tame.noop, __tame_fn_12, __tame_fn_17, tame.noop, __tame_fn_21, __tame_fn_25, __tame_k];
                if (__tame_switch_x == "zero") {
                    tame.callChain(__tame_switch_calls);
                } else {
                    __tame_switch_calls.shift();
                    if (__tame_switch_x == "seventeen") {
                        tame.callChain(__tame_switch_calls);
                    } else {
                        __tame_switch_calls.shift();
                        if (__tame_switch_x == "one") {
                            tame.callChain(__tame_switch_calls);
                        } else {
                            __tame_switch_calls.shift();
                            if (__tame_switch_x == "two") {
                                tame.callChain(__tame_switch_calls);
                            } else {
                                __tame_switch_calls.shift();
                                if (__tame_switch_x == "eighteen") {
                                    tame.callChain(__tame_switch_calls);
                                } else {
                                    __tame_switch_calls.shift();
                                    if (__tame_switch_x == "three") {
                                        tame.callChain(__tame_switch_calls);
                                    } else {
                                        __tame_switch_calls.shift();
                                        tame.callChain(__tame_switch_calls);
                                    }
                                }
                            }
                        }
                    }
                }
                tame.setActiveCb (null);
            };
            var __tame_fn_30 = function (__tame_k) {
                tame.setActiveCb (__tame_defer_cb);
                cb ( ) ;
                tame.callChain([__tame_k]);
                tame.setActiveCb (null);
            };
            tame.callChain([__tame_fn_6, __tame_fn_30, __tame_k]);
            tame.setActiveCb (null);
        };
        tame.callChain([__tame_fn_5, __tame_k]);
        tame.setActiveCb (null);
    }
    exports [ "switch & case statement" ] = {
    
    run :
    function  (cb) {
        var __tame_defer_cb = tame.findDeferCb ([cb]);
        tame.setActiveCb (__tame_defer_cb);
        var __tame_this = this;
        var __tame_fn_39 = function (__tame_k) {
            tame.setActiveCb (__tame_defer_cb);
            var lst = [ "zero" , "one" , "two" , "three" , "four" ,
            "eighteen" , "seventeen" ] ;
            var __tame_fn_31 = function (__tame_k) {
                tame.setActiveCb (__tame_defer_cb);
                var __tame_fn_32 = function (__tame_k) {
                    tame.setActiveCb (__tame_defer_cb);
                    var __tame_k_implicit = {};
                    var __tame_for_in_array = [];
                    for (var i in lst) { __tame_for_in_array.push(i); }
                    var __tame_for_in_iter = 0;
                    var __tame_fn_33 = function (__tame_k) {
                        tame.setActiveCb (__tame_defer_cb);
                        var __tame_fn_34 = function (__tame_k) {
                            tame.setActiveCb (__tame_defer_cb);
                            __tame_for_in_iter++
                            tame.callChain([__tame_fn_33, __tame_k]);
                            tame.setActiveCb (null);
                        };
                        __tame_k_implicit.k_break = __tame_k;
                        __tame_k_implicit.k_continue = function() { __tame_fn_34(__tame_k); };
                        if (__tame_for_in_iter < __tame_for_in_array.length) {
                            var __tame_fn_37 = function (__tame_k) {
                                tame.setActiveCb (__tame_defer_cb);
                                var i = __tame_for_in_array[__tame_for_in_iter];
                                var __tame_fn_35 = function (__tame_k) {
                                    tame.setActiveCb (__tame_defer_cb);
                                    var __tame_defers = new tame.Deferrals (__tame_k);
                                    var __tame_fn_36 = function (__tame_k) {
                                        tame.setActiveCb (__tame_defer_cb);
                                        runsw ( lst [ i ] ,
                                        __tame_defers.defer ( { 
                                            parent_cb : __tame_defer_cb,
                                            line : 47,
                                            file : "test/cases/24.tjs"
                                        } )
                                        ) ;
                                        tame.callChain([__tame_k]);
                                        tame.setActiveCb (null);
                                    };
                                    __tame_fn_36(tame.end);
                                    __tame_defers._fulfill();
                                    tame.setActiveCb (null);
                                };
                                tame.callChain([__tame_fn_35, __tame_k]);
                                tame.setActiveCb (null);
                            };
                            tame.callChain([__tame_fn_37, __tame_fn_34, __tame_k]);
                        } else {
                            tame.callChain([__tame_k]);
                        }
                        tame.setActiveCb (null);
                    };
                    tame.callChain([__tame_fn_33, __tame_k]);
                    tame.setActiveCb (null);
                };
                var __tame_fn_38 = function (__tame_k) {
                    tame.setActiveCb (__tame_defer_cb);
                    cb ( ) ;
                    tame.callChain([__tame_k]);
                    tame.setActiveCb (null);
                };
                tame.callChain([__tame_fn_32, __tame_fn_38, __tame_k]);
                tame.setActiveCb (null);
            };
            tame.callChain([__tame_fn_31, __tame_k]);
            tame.setActiveCb (null);
        };
        tame.callChain([__tame_fn_39, __tame_k]);
        tame.setActiveCb (null);
    }
    ,
    
    check :
    function  () {
        assert . deepEqual ( results , [ 0 , 1 , 2 , 3 , 4 , 3 , 4 , 4 , 3 , 4 , 1 ] , "cascade check" ) ;
    }
    } ;
    tame.callChain([__tame_k]);
    tame.setActiveCb (null);
};
__tame_fn_0 (tame.end);
