// tamejs/test/cases/35.out.js
var tame = require('tamejs').runtime;
var __tame_defer_cb = null;
var __tame_fn_0 = function (__tame_k) {
    tame.setActiveCb (__tame_defer_cb);
    "use strict" ;
    
    require ( 'tamejs' ) . register ( ) ;
    var assert = require ( "assert" ) ;
    var Pipeliner = require ( '../../lib/connectors.tjs' ) . Pipeliner ;
    var slots = { } ;
    
    exports [ "pipelining" ] = {
    
    
    run :
    function  (ev) {
        var __tame_defer_cb = tame.findDeferCb ([ev]);
        tame.setActiveCb (__tame_defer_cb);
        var __tame_this = this;
        var __tame_fn_5 = function (__tame_k) {
            tame.setActiveCb (__tame_defer_cb);
            function call (i, cb) {
                var __tame_defer_cb = tame.findDeferCb ([i, cb]);
                tame.setActiveCb (__tame_defer_cb);
                var __tame_this = this;
                var __tame_fn_10 = function (__tame_k) {
                    tame.setActiveCb (__tame_defer_cb);
                    slots [ i ] = 1 ;
                    var __tame_fn_6 = function (__tame_k) {
                        tame.setActiveCb (__tame_defer_cb);
                        var __tame_fn_7 = function (__tame_k) {
                            tame.setActiveCb (__tame_defer_cb);
                            var __tame_defers = new tame.Deferrals (__tame_k);
                            var __tame_fn_8 = function (__tame_k) {
                                tame.setActiveCb (__tame_defer_cb);
                                setTimeout (
                                __tame_defers.defer ( { 
                                    func_name : "call",
                                    parent_cb : __tame_defer_cb,
                                    line : 17,
                                    file : "test/cases/35.tjs"
                                } )
                                , 10*Math . random ( ) ) ;
                                tame.callChain([__tame_k]);
                                tame.setActiveCb (null);
                            };
                            __tame_fn_8(tame.end);
                            __tame_defers._fulfill();
                            tame.setActiveCb (null);
                        };
                        var __tame_fn_9 = function (__tame_k) {
                            tame.setActiveCb (__tame_defer_cb);
                            slots [ i ] |= 2 ;
                            cb ( 4 ) ;
                            tame.callChain([__tame_k]);
                            tame.setActiveCb (null);
                        };
                        tame.callChain([__tame_fn_7, __tame_fn_9, __tame_k]);
                        tame.setActiveCb (null);
                    };
                    tame.callChain([__tame_fn_6, __tame_k]);
                    tame.setActiveCb (null);
                };
                tame.callChain([__tame_fn_10, __tame_k]);
                tame.setActiveCb (null);
            }
            ;
            function window (n, w, cb) {
                var __tame_defer_cb = tame.findDeferCb ([n, w, cb]);
                tame.setActiveCb (__tame_defer_cb);
                var __tame_this = this;
                var __tame_fn_23 = function (__tame_k) {
                    tame.setActiveCb (__tame_defer_cb);
                    var pipeliner = new Pipeliner ( w , .01 ) ;
                    var tmp = { } ;
                    var __tame_fn_11 = function (__tame_k) {
                        tame.setActiveCb (__tame_defer_cb);
                        var __tame_fn_12 = function (__tame_k) {
                            tame.setActiveCb (__tame_defer_cb);
                            var __tame_k_implicit = {};
                            var i = 0;
                            var __tame_fn_13 = function (__tame_k) {
                                tame.setActiveCb (__tame_defer_cb);
                                var __tame_fn_14 = function (__tame_k) {
                                    tame.setActiveCb (__tame_defer_cb);
                                    i ++
                                    tame.callChain([__tame_fn_13, __tame_k]);
                                    tame.setActiveCb (null);
                                };
                                __tame_k_implicit.k_break = __tame_k;
                                __tame_k_implicit.k_continue = function() { __tame_fn_14(__tame_k); };
                                if (i < n) {
                                    var __tame_fn_15 = function (__tame_k) {
                                        tame.setActiveCb (__tame_defer_cb);
                                        var __tame_fn_16 = function (__tame_k) {
                                            tame.setActiveCb (__tame_defer_cb);
                                            var __tame_defers = new tame.Deferrals (__tame_k);
                                            var __tame_fn_17 = function (__tame_k) {
                                                tame.setActiveCb (__tame_defer_cb);
                                                pipeliner . waitInQueue (
                                                __tame_defers.defer ( { 
                                                    func_name : "window",
                                                    parent_cb : __tame_defer_cb,
                                                    line : 26,
                                                    file : "test/cases/35.tjs"
                                                } )
                                                ) ;
                                                tame.callChain([__tame_k]);
                                                tame.setActiveCb (null);
                                            };
                                            __tame_fn_17(tame.end);
                                            __tame_defers._fulfill();
                                            tame.setActiveCb (null);
                                        };
                                        var __tame_fn_18 = function (__tame_k) {
                                            tame.setActiveCb (__tame_defer_cb);
                                            call ( i , pipeliner .
                                            __tame_defers.defer ( { 
                                                assign_fn : 
                                                    (function (__tame_slot_0, __tame_slot_1) {
                                                        return function () { 
                                                            __tame_slot_0[__tame_slot_1] = arguments[0];
                                                        }
                                                        ;
                                                    }) (tmp, i),
                                                func_name : "window",
                                                parent_cb : __tame_defer_cb,
                                                line : 27,
                                                file : "test/cases/35.tjs"
                                            } )
                                            ) ;
                                            tame.callChain([__tame_k]);
                                            tame.setActiveCb (null);
                                        };
                                        tame.callChain([__tame_fn_16, __tame_fn_18, __tame_k]);
                                        tame.setActiveCb (null);
                                    };
                                    tame.callChain([__tame_fn_15, __tame_fn_14, __tame_k]);
                                } else {
                                    tame.callChain([__tame_k]);
                                }
                                tame.setActiveCb (null);
                            };
                            tame.callChain([__tame_fn_13, __tame_k]);
                            tame.setActiveCb (null);
                        };
                        var __tame_fn_19 = function (__tame_k) {
                            tame.setActiveCb (__tame_defer_cb);
                            var __tame_defers = new tame.Deferrals (__tame_k);
                            var __tame_fn_20 = function (__tame_k) {
                                tame.setActiveCb (__tame_defer_cb);
                                pipeliner . flush (
                                __tame_defers.defer ( { 
                                    func_name : "window",
                                    parent_cb : __tame_defer_cb,
                                    line : 29,
                                    file : "test/cases/35.tjs"
                                } )
                                ) ;
                                tame.callChain([__tame_k]);
                                tame.setActiveCb (null);
                            };
                            __tame_fn_20(tame.end);
                            __tame_defers._fulfill();
                            tame.setActiveCb (null);
                        };
                        var __tame_fn_21 = function (__tame_k) {
                            tame.setActiveCb (__tame_defer_cb);
                             for (var i in tmp) {
                                slots [ i ] |= tmp [ i ] ;
                            }
                            tame.callChain([__tame_k]);
                            tame.setActiveCb (null);
                        };
                        var __tame_fn_22 = function (__tame_k) {
                            tame.setActiveCb (__tame_defer_cb);
                            cb ( ) ;
                            tame.callChain([__tame_k]);
                            tame.setActiveCb (null);
                        };
                        tame.callChain([__tame_fn_12, __tame_fn_19, __tame_fn_21, __tame_fn_22, __tame_k]);
                        tame.setActiveCb (null);
                    };
                    tame.callChain([__tame_fn_11, __tame_k]);
                    tame.setActiveCb (null);
                };
                tame.callChain([__tame_fn_23, __tame_k]);
                tame.setActiveCb (null);
            }
            ;
            
            var n = 100 ;
            var __tame_fn_1 = function (__tame_k) {
                tame.setActiveCb (__tame_defer_cb);
                var __tame_fn_2 = function (__tame_k) {
                    tame.setActiveCb (__tame_defer_cb);
                    var __tame_defers = new tame.Deferrals (__tame_k);
                    var __tame_fn_3 = function (__tame_k) {
                        tame.setActiveCb (__tame_defer_cb);
                        window ( n , 10 ,
                        __tame_defers.defer ( { 
                            parent_cb : __tame_defer_cb,
                            line : 37,
                            file : "test/cases/35.tjs"
                        } )
                        ) ;
                        tame.callChain([__tame_k]);
                        tame.setActiveCb (null);
                    };
                    __tame_fn_3(tame.end);
                    __tame_defers._fulfill();
                    tame.setActiveCb (null);
                };
                var __tame_fn_4 = function (__tame_k) {
                    tame.setActiveCb (__tame_defer_cb);
                    ev ( ) ;
                    tame.callChain([__tame_k]);
                    tame.setActiveCb (null);
                };
                tame.callChain([__tame_fn_2, __tame_fn_4, __tame_k]);
                tame.setActiveCb (null);
            };
            tame.callChain([__tame_fn_1, __tame_k]);
            tame.setActiveCb (null);
        };
        tame.callChain([__tame_fn_5, __tame_k]);
        tame.setActiveCb (null);
    }
    ,
    
    check :
    function  () {
         for (var i in slots) {
            assert . equal ( slots [ i ] , 7 , "all 3 events make bitwise 7" ) ;
        }
    }
    } ;
    tame.callChain([__tame_k]);
    tame.setActiveCb (null);
};
__tame_fn_0 (tame.end);
